### PR TITLE
Circuit dynamic configure

### PR DESCRIPTION
**Pull request by freeranged3v**
Credit: https://github.com/privacy-scaling-explorations/halo2/pull/168

### CODE REVIEW

File: halo2_proofs/src/lib.rs [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #![deny(missing_debug_implementations)]
 #![deny(missing_docs)]
 #![deny(unsafe_code)]
+#![feature(associated_type_defaults)]
```
Comment (freeranged3v) [inline halo2_proofs/src/lib.rs:10]: Unstable feature here.

File: halo2_proofs/src/plonk/circuit.rs [hunk 2]
```diff
@@ -470,14 +470,34 @@ pub trait Circuit<F: Field> {
     /// `Circuit` trait because its behaviour is circuit-critical.
     type FloorPlanner: FloorPlanner;
 
+    /// Runtime parameters to configure the circuit.
+    #[cfg(feature = "circuit-params")]
+    type Params: Default = ();
```
Comment (freeranged3v) [inline halo2_proofs/src/plonk/circuit.rs:475]: Associated type default: https://github.com/rust-lang/rust/issues/29661

File: halo2_proofs/src/plonk/prover.rs [hunk 3]
```diff
@@ -724,63 +728,128 @@ pub fn create_proof<
...
+    {
+        layouter.assign_region(
+            || "load private",
+            |mut region| region.assign_advice(|| "load private", column, 0, || value),
+        )
+    }
 
-    impl<F: Field> Circuit<F> for MyCircuit {
-        type Config = ();
+    impl Circuit<pallas::Base> for MyCircuit {
```
Comment (freeranged3v) [inline halo2_proofs/src/plonk/prover.rs:768]: Test case that should be helpful to see the flow.

File: halo2_proofs/Cargo.toml [hunk 4]
```diff
@@ -54,7 +54,7 @@ blake2b_simd = "1"
 maybe-rayon = {version = "0.1.0", default-features = false}
 
 # Developer tooling dependencies
-plotters = { version = "0.3.0", default-features = false, optional = true }
+plotters = { version = "0.3.0", optional = true }
```
Comment (freeranged3v) [inline halo2_proofs/Cargo.toml:57]: I turned on the default features to generate a test circuit layout, the PSE fork has the defaults on too.

File: halo2_proofs/src/plonk/circuit.rs (hunk 2 above)
Comment (freeranged3v) [reply halo2_proofs/src/plonk/circuit.rs:475]: @parazyd Would turning on this unstable feature - associated type default - be a good idea? 

I think tradeoff is:

the pro:
- won't break any circuits implementing the Circuit trait from zcash/halo2, which seems nice, one can use zcash/halo2 `Circuit` implementing circuits with this fork directly

the cons:
- potential corner cases (though the usage is simple here)
- need to migrate if and when Rust team introduces a breaking change (but the alternative is breaking the circuits now)

other misc points:
- Using an unstable feature requires depending on `nightly` which we already do in CI, so I think it isn't an issue.

**Comment by parazyd** [comment]
Thanks a lot. I'll merge this manually and add a few tweaks.

**Pull request closed without merging**